### PR TITLE
fix: allow for base class

### PR DESCRIPTION
**Pull request by m4thieulavoie**
** Discussion PR, the goal is to get traction and kick off a discussion here **

# Pull Request

## 📖 Description

This PR adds the ability to specify the `baseClass` of a newly created FAST component. This is useful when using a typescript mixin on top of FAST foundation components. Since under the hood FAST is looking for the exact base class name, this allows to specify and override the default value.

```
class MyCustomSwitch extends Switch {}

export const fastSwitch = MyCustomSwitch.compose<SwitchOptions>({
    ...,
    baseClass: Switch
});
```

### 🎫 Issues

- https://github.com/microsoft/fast/issues/4941

### General

- [x] I have included a change request file using `$ yarn change`
- [ ] I have added tests for my changes.
- [ ] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

- [ ] I have added a new component
- [ ] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

### CODE REVIEW

File: packages/web-components/fast-foundation/src/foundation-element/foundation-element.ts
Comment (EisenbergEffect) [inline packages/web-components/fast-foundation/src/foundation-element/foundation-element.ts:224]: I *think* that passing the base class here instead of the actual class may cause some problems with the platform element definition. What if we allow an optional third argument to `tryDefineElement` that takes an `options` object that allows passing in an additional type for name mapping?

Comment (m4thieulavoie) [reply packages/web-components/fast-foundation/src/foundation-element/foundation-element.ts:224]: I think you're right, and either way it would be a lot safer to implement, I'll add it! 👍

Comment (m4thieulavoie) [reply packages/web-components/fast-foundation/src/foundation-element/foundation-element.ts]: @EisenbergEffect I ended up breaking the parameter list for an object for scalability, let me know what you think about it. I also chose to register the actual class _and_ the baseClass (if any) for better flexibility. I could really use your review at this point just to make sure what I'm doing still makes sense 😅

**Comment by m4thieulavoie** [comment]
@EisenbergEffect got the chance to have a look? I would love to hear your initial thoughts on this one

**Comment by EisenbergEffect** [comment]
I've got it on my list. It's looking like next week though due to my current commitments. I haven't forgotten it though and want to get a solution in.

File: packages/web-components/fast-foundation/docs/api-report.md
Comment (EisenbergEffect) [inline packages/web-components/fast-foundation/docs/api-report.md:1019]: Should we also enable an array? Do we like `baseClass` or would we prefer something like `alias` or `foundation`?

File: packages/web-components/fast-foundation/src/design-system/design-system.ts
Comment (EisenbergEffect) [inline packages/web-components/fast-foundation/src/design-system/design-system.ts:79]: I think we should probably export this and then add documentation comments. We'll also need the `@public` indicator in the comment to satisfy our docs generation stuff.

File: packages/web-components/fast-foundation/src/design-system/design-system.ts
Comment (EisenbergEffect) [inline packages/web-components/fast-foundation/src/design-system/design-system.ts:303]: Yep. I think it's probably this simple and should address the problem. I think this will help with a couple of Fluent/FAST components as well that need an additional registration.

File: packages/web-components/fast-foundation/src/foundation-element/foundation-element.ts
Comment (EisenbergEffect) [inline packages/web-components/fast-foundation/src/foundation-element/foundation-element.ts:37]:
```suggestion
     * The actual FAST base class of the component if different from the class used to compose.
```

File: packages/web-components/fast-foundation/src/design-system/design-system.ts
Comment (EisenbergEffect) [inline packages/web-components/fast-foundation/src/design-system/design-system.ts:104]: Though not used directly by consumers (as far as I know), this would technically be a breaking change. Do we feel like we can get away with that on a minor release? /cc @chrisdholt

Comment (EisenbergEffect) [reply packages/web-components/fast-foundation/src/design-system/design-system.ts:104]: Note: This is a better design as it would be more open to extension in the future without breaks. Just want to make sure that we're ok doing this on a minor release. I seriously doubt it will affect anyone, but again, technical breaking.

**Review by EisenbergEffect — CHANGES_REQUESTED** [review]
Sorry it took me so long to dig in here. I definitely want to address this issue with a solution very close to what's here, so I think this can move out of a Draft state now. I've noted a few things we should discuss before merging.  Also, before merging, we'll want at least one unit test that checks this functionality. If you aren't sure how to write that, let me know and I can probably jump in and add a commit with that after the rest of the issues are addressed.

File: packages/web-components/fast-foundation/docs/api-report.md
Comment (m4thieulavoie) [reply packages/web-components/fast-foundation/docs/api-report.md:1019]: Are you proposing doing something like the following?
```suggestion
    baseClass?: Constructable | Constructable[];
```

For the name, I'll leave it up to you guys. However, `foundation` sounds a bit vague to me

Comment (EisenbergEffect) [reply packages/web-components/fast-foundation/docs/api-report.md:1019]: I agree. I think that was a bad idea. We'll probably keep this as `baseClass` and if we need we can make it single or array, but we can add the array support in a later PR if someone really needs it.

Comment (m4thieulavoie) [reply packages/web-components/fast-foundation/docs/api-report.md:1019]: Sounds good with me!

File: packages/web-components/fast-foundation/src/design-system/design-system.ts
Comment (m4thieulavoie) [reply packages/web-components/fast-foundation/src/design-system/design-system.ts]: Should be fixed now 👍

**Comment by EisenbergEffect** [comment]
@m4thieulavoie Ready for me to take another look?

**Comment by m4thieulavoie** [comment]
> @m4thieulavoie Ready for me to take another look?

Yes! That's why I requested back your review with the GitHub feature a couple of days ago 👍

**Comment by EisenbergEffect** [comment]
Oops! Sorry about that! A lot of emails fly through my inbox...I may have missed the alert for that one. It's on my list to review now 😄

File: packages/web-components/fast-foundation/src/design-system/design-system.ts
Comment (EisenbergEffect) [inline packages/web-components/fast-foundation/src/design-system/design-system.ts:111]: I think we need to fix this param description since this is technically not the definition context.

File: packages/web-components/fast-foundation/src/foundation-element/foundation-element.ts
Comment (EisenbergEffect) [inline packages/web-components/fast-foundation/src/foundation-element/foundation-element.ts:224]: Should we Omit the `baseClass` from the `OverrideFoundationElementDefinition` as well?

**Review by EisenbergEffect — CHANGES_REQUESTED** [review]
Almost there. One thing we need to fix in the docs comments and then a question I have for consideration.

**Comment by m4thieulavoie** [comment]
> Almost there. One thing we need to fix in the docs comments and then a question I have for consideration.

Thanks! I'm AFK until monday but I'll correct those once I'm back!

File: packages/web-components/fast-foundation/src/design-system/design-system.ts
Comment (m4thieulavoie) [reply packages/web-components/fast-foundation/src/design-system/design-system.ts:111]: I had a bit of a hard time finding a proper description for this one. I could use some suggestions, otherwise I'll come up with something on monday 👍

Comment (vaudy) [reply packages/web-components/fast-foundation/src/design-system/design-system.ts:111]: Maybe just using `Element Definition`? @EisenbergEffect any idea?

File: packages/web-components/fast-foundation/src/foundation-element/foundation-element.ts
Comment (m4thieulavoie) [reply packages/web-components/fast-foundation/src/foundation-element/foundation-element.ts:224]: I'm not sure I follow the question 🤔

Comment (m4thieulavoie) [reply packages/web-components/fast-foundation/src/foundation-element/foundation-element.ts:224]: Nevermind, just got the sense of the question (I need cafeine 😅 ). I think so yeah!

Comment (m4thieulavoie) [reply packages/web-components/fast-foundation/src/foundation-element/foundation-element.ts:224]: Actually we can't `omit` it. `overrideDefinition` has the `OverrideFoundationElementDefinition` type
image

File: packages/web-components/fast-foundation/src/design-system/design-system.ts
Comment (m4thieulavoie) [reply packages/web-components/fast-foundation/src/design-system/design-system.ts]: I highly inspired from @vaudy 's suggestion, but feel free to suggest anything else 👍

File: packages/web-components/fast-foundation/src/foundation-element/foundation-element.ts
Comment (EisenbergEffect) [reply packages/web-components/fast-foundation/src/foundation-element/foundation-element.ts:224]: I'm a bit confused because the code below does exactly this:

```ts
baseClass: overrideDefinition.baseClass
```

But it seems like that should never come from the override definition but only from the original definition.

Comment (m4thieulavoie) [reply packages/web-components/fast-foundation/src/foundation-element/foundation-element.ts:224]: I think I see what you mean. The line in your previous comment is required as `baseClass` is passed along _twice_ in order to get from the `compose()` call all the way to the actual condition added in the other file changed in this PR.
I hope I was clear! :) However if I missed something definitely let me know!

**Comment by EisenbergEffect** [comment]
@nicholasrice Could you take another look at this? I think @m4thieulavoie has addressed all the issues.

File: packages/web-components/fast-foundation/src/design-system/design-system.ts
Comment (nicholasrice) [reply packages/web-components/fast-foundation/src/design-system/design-system.ts]: yep, you can deprecate overload signatures :)

**Review by nicholasrice — APPROVED** [review]

**Comment by m4thieulavoie** [comment]
@EisenbergEffect @nicholasrice I had a pretty big merge conflict to fix, so please make sure to double-check the changes before merging in in case I messed up something 👍

**Review by EisenbergEffect — CHANGES_REQUESTED** [review]
Unfortunately, I don't think the tests for the improved disambiguation are correct. Can you update those from what's in our main branch?

**Comment by m4thieulavoie** [comment]
> Unfortunately, I don't think the tests for the improved disambiguation are correct. Can you update those from what's in our main branch?

I think I would need more guidance on this, they seem to be correct as far as I can tell. Can you maybe point out the tests are are erroned @EisenbergEffect ?

**Comment by EisenbergEffect** [comment]
Some of the assertions are missing and also the setting of `callbackCalled` in one of the tests. Here's the source to replace the two incorrect tests:

```ts
    it("Should only call callbacks for duplicate elements by default", () => {
        const elementName = uniqueElementName();
        const customElement = class extends HTMLElement {};
        const host = document.createElement("div");
        const system = DesignSystem.getOrCreate(host);
        let callbackCalled = false;

        expect(() => {
            system.register(
                {
                    register(container: Container) {
                        const context = container.get(DesignSystemRegistrationContext);
                        context.tryDefineElement(elementName, customElement, x =>
                            x.defineElement()
                        );
                    },
                },
                {
                    register(container: Container) {
                        const context = container.get(DesignSystemRegistrationContext);
                        context.tryDefineElement(
                            elementName,
                            class extends HTMLElement {},
                            x => {
                                x.defineElement();
                                callbackCalled = true;
                            }
                        );
                    },
                }
            );
        }).not.to.throw();

        expect(customElements.get(elementName)).to.equal(customElement);
        expect(callbackCalled).to.be.true;
    });

    it("Can completely ignore duplicates", () => {
        const elementName = uniqueElementName();
        const customElement = class extends HTMLElement {};
        const host = document.createElement("div");
        const system = DesignSystem.getOrCreate(host)
            .withElementDisambiguation(() => ElementDisambiguation.ignoreDuplicate);
        let callbackCalled = false;

        expect(() => {
            system.register(
                {
                    register(container: Container) {
                        const context = container.get(DesignSystemRegistrationContext);
                        context.tryDefineElement(elementName, customElement, x =>
                            x.defineElement()
                        );
                    },
                },
                {
                    register(container: Container) {
                        const context = container.get(DesignSystemRegistrationContext);
                        context.tryDefineElement(
                            elementName,
                            class extends HTMLElement {},
                            x => {
                                x.defineElement();
                                callbackCalled = true;
                            }
                        );
                    },
                }
            );
        }).not.to.throw();

        expect(customElements.get(elementName)).to.equal(customElement);
        expect(callbackCalled).to.be.false;
    });
```

**Comment by m4thieulavoie** [comment]
> Some of the assertions are missing and also the setting of `callbackCalled` in one of the tests. Here's the source to replace the two incorrect tests:
> 
> ```ts
>     it("Should only call callbacks for duplicate elements by default", () => {
>         const elementName = uniqueElementName();
>         const customElement = class extends HTMLElement {};
>         const host = document.createElement("div");
>         const system = DesignSystem.getOrCreate(host);
>         let callbackCalled = false;
> 
>         expect(() => {
>             system.register(
>                 {
>                     register(container: Container) {
>                         const context = container.get(DesignSystemRegistrationContext);
>                         context.tryDefineElement(elementName, customElement, x =>
>                             x.defineElement()
>                         );
>                     },
>                 },
>                 {
>                     register(container: Container) {
>                         const context = container.get(DesignSystemRegistrationContext);
>                         context.tryDefineElement(
>                             elementName,
>                             class extends HTMLElement {},
>                             x => {
>                                 x.defineElement();
>                                 callbackCalled = true;
>                             }
>                         );
>                     },
>                 }
>             );
>         }).not.to.throw();
> 
>         expect(customElements.get(elementName)).to.equal(customElement);
>         expect(callbackCalled).to.be.true;
>     });
> 
>     it("Can completely ignore duplicates", () => {
>         const elementName = uniqueElementName();
>         const customElement = class extends HTMLElement {};
>         const host = document.createElement("div");
>         const system = DesignSystem.getOrCreate(host)
>             .withElementDisambiguation(() => ElementDisambiguation.ignoreDuplicate);
>         let callbackCalled = false;
> 
>         expect(() => {
>             system.register(
>                 {
>                     register(container: Container) {
>                         const context = container.get(DesignSystemRegistrationContext);
>                         context.tryDefineElement(elementName, customElement, x =>
>                             x.defineElement()
>                         );
>                     },
>                 },
>                 {
>                     register(container: Container) {
>                         const context = container.get(DesignSystemRegistrationContext);
>                         context.tryDefineElement(
>                             elementName,
>                             class extends HTMLElement {},
>                             x => {
>                                 x.defineElement();
>                                 callbackCalled = true;
>                             }
>                         );
>                     },
>                 }
>             );
>         }).not.to.throw();
> 
>         expect(customElements.get(elementName)).to.equal(customElement);
>         expect(callbackCalled).to.be.false;
>     });
> ```

Oh right, GH made it really hard to compare, it should be fixed now 👍

**Review by EisenbergEffect — APPROVED** [review]

**Comment by EisenbergEffect** [comment]
Awesome! Thanks @m4thieulavoie We'll get this in as soon as CI finishes up.

**Comment by EisenbergEffect** [comment]
It's in! This should go out in this evening's release.

**Pull request merged**
